src/Xamarin.Android.Build.Tasks/Tasks/GenerateJavaStubs.cs

```diff
@@ -183,26 +185,36 @@ void Run (bool useMarshalMethods)
 			}
 
 			// Now that "never" never happened, we can proceed knowing that at least the assembly sets are the same for each architecture
-			var nativeCodeGenStates = new Dictionary<AndroidTargetArch, NativeCodeGenState> ();
-			bool generateJavaCode = true;
+			var nativeCodeGenStates = new ConcurrentDictionary<AndroidTargetArch, NativeCodeGenState> ();
 			NativeCodeGenState? templateCodeGenState = null;
 
-			foreach (var kvp in allAssembliesPerArch) {
+			var firstArch = allAssembliesPerArch.First ().Key;
+			var generateSucceeded = true;
+
+			// Generate Java sources in parallel
+			Parallel.ForEach (allAssembliesPerArch, (kvp) => {
 				AndroidTargetArch arch = kvp.Key;
 				Dictionary<string, ITaskItem> archAssemblies = kvp.Value;
+
+				// We only need to generate Java code for one ABI, as the Java code is ABI-agnostic
+				// Pick the "first" one as the one to generate Java code for
+				var generateJavaCode = arch == firstArch;
+
 				(bool success, NativeCodeGenState? state) = GenerateJavaSourcesAndMaybeClassifyMarshalMethods (arch, archAssemblies, MaybeGetArchAssemblies (userAssembliesPerArch, arch), useMarshalMethods, generateJavaCode);
 
-				if (!success) {
-					return;
-				}
+				if (!success)
+					generateSucceeded = false;
```

Comment:
One nitpick, could we use curly braces around one-liner statements like here?

Reply:
I can, but it violates our stated coding standards, so I'm not sure if we want to:
https://www.mono-project.com/community/contributing/coding-guidelines/#where-to-put-braces

Reply:
Standards are a guide ;-)

I'm with @grendello, and prefer using braces "everywhere", even if they'd otherwise be optional.

(Notable exception: when using `=>` expression bodied members for constructors, properties, methods, etc.)

Reply:
Updated.

Reply:
@jpobst to put it into perspective why braces might be very important, see https://www.synopsys.com/blogs/software-security/understanding-apple-goto-fail-vulnerability-2.html